src/extension/index.ts

```diff
@@ -109,6 +111,52 @@ function activateDiagnostics(disposables: Disposable[], store: Store, baser: Uri
     disposables.push(workspace.onDidCloseTextDocument(doc => diagsAll.delete(doc.uri))); // Spurious *.git deletes don't hurt.
     const disposer = observe(store.logs, () => workspace.textDocuments.forEach(setDiags));
     disposables.push({ dispose: disposer });
+    return setDiags;
+}
+
+function activateCodeActions(disposables: Disposable[], baser: UriRebaser, refreshDiagnostics: (document: TextDocument) => void) {
+    const commandId = 'sarif.fix';
+    disposables.push(commands.registerCommand(commandId, async (document: TextDocument, result: Result, fix: Fix) => {
+        const edit = new WorkspaceEdit();
+        for (const change of fix.artifactChanges) {
+            const { artifactLocation, replacements } = change;
+            const [artifactUri] = parseArtifactLocation(result, artifactLocation);
+            if (!artifactUri) continue;
+
+            const localUri = await baser.translateArtifactToLocal(artifactUri);
+            for (const replacement of replacements) {
+                edit.replace(
+                    Uri.parse(localUri),
+                    regionToSelection(document, replacement.deletedRegion),
+                    replacement.insertedContent?.text ?? ''
+                );
+            }
+        }
+        workspace.applyEdit(edit);
+        result._fixed = true;
+        refreshDiagnostics(document);
+    }));
+
+    const provider = new class implements CodeActionProvider {
```

Comment:
>CodeActionProvider [](start = 42, length = 18)

So why not move this implementation to its own file?

Reply:
Discussed offline. For other reviewers, see first thread above about "2 functions nested".